### DISCUSSION TITLE
Issues with form success/error messages when templates are cached

### DISCUSSION THREAD

**Discussion by jrrdnx** [solspace/craft-freeform#280] [Questions & Answers] [answered]
I've followed the directions at https://docs.solspace.com/craft/freeform/v3/templating/caching.html#twig-template-caching to update the form's hash input and csrf input. This is working correctly because form submissions are actually coming through.

However, no success message or error message is displayed when the form is submitted. It appears as if nothing has happened when the result page is displayed, even if the submission was successful (email notification is received and submission appears on the back-end).

If the template cache has been cleared and the form submitted immediately (without reloading first to establish the cache), then the resulting success/error message is then cached so they'll appear on every page load regardless of if the form was even submitted.

I am using the honeypot feature, but am NOT using the JS enhancement.

I did have to slightly modify the example code, as we are using Forms within a Neo fieldtype to allow clients to add forms to pages at their discretion:

```twig
<section data-form="{{ count }}">
  {% set form = craft.freeform.form(form.handle) %}

  {% cache %}
    {{ form.render }}
  {% endcache %}

  <script>
    // Find the corresponding Form
    var form = document.querySelector('[data-form="{{ count }}"] form');

    // Locate and update the Hash input
    var formHashInput = form.querySelector('input[name=formHash]');
    formHashInput.setAttribute('value', '{{ form.hash }}');

    // Locate and update the CSRF input
    var csrfInput = form.querySelector('input[name={{ craft.app.config.general.csrfTokenName|e('js') }}]');
    csrfInput.value = '{{ craft.app.request.csrfToken|e('js') }}';
  </script>
</section>
```

**Comment by kjmartens** [comment, marked as answer]
Hi @jrrdnx, are you using the built-in AJAX feature with this form as well? If you're wanting that success message to display while keeping the user on the same page, you'll need to use AJAX, otherwise, you'll experience the issue you describe. 🙂

**Reply by jrrdnx** [reply]
Not with this particular form, no. We leave this open (with proper training) for clients to determine if they want to use AJAX or a different Return URL, so can't always assume this will be enabled. Thanks!
